### PR TITLE
Add override for binary_cross_entropy and its backward.

### LINKED ISSUE #1271: Change the "reduction" of Pytorch Losses
## 🚀 Feature
 Many torch loss functions, e.g., cross entropy, nll loss, etc., have an ability to change the `reduce` from mean to other values. This is used in multiple places inside Fairseq, where we would like to not take the mean and instead use the sum or not reduce the loss at all. This is especiallyuseful for changing the normalization / doing masking. For example, here https://github.com/pytorch/fairseq/blob/master/fairseq/criterions/masked_lm.py#L54

Currently, the loss functions are supported in XLA, but setting the reduce flags is unsupported and the objects move to the CPU (i.e., I see `aten:` calls for the cross entropy). This slows speeds significantly 

In the meantime, if your end goal is to have a `sum` instead of a `mean`, you can simply call the regular loss and then multiply by the number of elements. However, there are also cases where we need `none` reduction, which there is no work around that I am aware of.

### PR DESCRIPTION
**Pull request by ailzhang**
fixes #1271

### CODE REVIEW

File: torch_xla/csrc/aten_xla_type.cpp [hunk 1]
```diff
@@ -655,6 +655,61 @@ at::Tensor& AtenXlaType::bernoulli_(at::Tensor& self, const at::Tensor& p,
...
+  XLA_FN_COUNTER("xla::");
+  XLATensor self_tensor = bridge::GetXlaTensor(self);
+  XLATensor weight_tensor =
+      bridge::GetOrCreateXlaTensor(weight, self_tensor.GetDevice());
+  return bridge::AtenFromXlaTensor(XLATensor::binary_cross_entropy_backward(
+      bridge::GetXlaTensor(grad_output), self_tensor,
+      bridge::GetXlaTensor(target), weight_tensor, reduction));
+}
+
+at::Tensor& AtenXlaType::binary_cross_entropy_backward_out(
```
Comment (dlibenzi) [inline torch_xla/csrc/aten_xla_type.cpp]: We cannot move forward and have two implementations of every operator 😐 
Who is calling the out function in these cases?
Otherwise we need to put in place something which allow us to implement only the _out one, and route the other one to the _out one.

File: torch_xla/csrc/ops/binary_cross_entropy_backward.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    const absl::optional<Value>& weight, ReductionMode reduction)
+    : Node(ir::OpKind(at::aten::binary_cross_entropy_backward),
+           xla::util::GetValuesVector<Value>({grad_output, logits, labels},
+                                             {&weight}),
+           [&]() {
+             return NodeOutputShape(grad_output, logits, labels, weight,
+                                    reduction);
+           },
+           /*num_outputs=*/1,
+           xla::util::MHash(xla::util::GetEnumValue<ReductionMode>(reduction))),
```
Comment (dlibenzi) [inline torch_xla/csrc/ops/binary_cross_entropy_backward.cpp]: No need for template specification.

File: torch_xla/csrc/ops/binary_cross_entropy.cpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+}  // namespace
+
+BinaryCrossEntropy::BinaryCrossEntropy(const Value& logits, const Value& labels,
+                                       const absl::optional<Value>& weight,
+                                       ReductionMode reduction)
+    : Node(ir::OpKind(at::aten::binary_cross_entropy),
+           xla::util::GetValuesVector<Value>({logits, labels}, {&weight}),
+           [&]() { return NodeOutputShape(logits, labels, weight, reduction); },
+           /*num_outputs=*/1,
+           xla::util::MHash(xla::util::GetEnumValue<ReductionMode>(reduction))),
```
Comment (dlibenzi) [inline torch_xla/csrc/ops/binary_cross_entropy.cpp]: No need for template specification. Posting a PR to remove it from other places as well.

File: torch_xla/csrc/ops/binary_cross_entropy.cpp [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    weight = loctx->GetOutputOp(operand(2));
+  }
+  return ReturnOp(BuildBinaryCrossEntropy(logits, labels, weight, reduction_),
+                  loctx);
+}
+
+std::string BinaryCrossEntropy::ToString() const {
+  std::stringstream ss;
+  ss << Node::ToString()
+     << ", reduction=" << xla::util::GetEnumValue<ReductionMode>(reduction_);
```
Comment (dlibenzi) [inline torch_xla/csrc/ops/binary_cross_entropy.cpp]: No need for template specification.

File: torch_xla/csrc/ops/binary_cross_entropy_backward.cpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  }
+  return ReturnOp(BuildBinaryCrossEntropyBackward(grad_output, logits, labels,
+                                                  weight, reduction_),
+                  loctx);
+}
+
+std::string BinaryCrossEntropyBackward::ToString() const {
+  std::stringstream ss;
+  ss << Node::ToString()
+     << ", reduction=" << xla::util::GetEnumValue<ReductionMode>(reduction_);
```
Comment (dlibenzi) [inline torch_xla/csrc/ops/binary_cross_entropy_backward.cpp]: No need for template specification.

File: torch_xla/csrc/aten_xla_type.cpp (hunk 1 above)
Comment (ailzhang) [reply torch_xla/csrc/aten_xla_type.cpp]: @dlibenzi In this case it's special because both `bce` and its `out` version are still in our legacy THNN code. So both of them are implemented through one implementation. When they're moved into aten in the future, I think the `out` version will be kept as leaf. But currently when both `out` and normal version can be overrided, overriding normal ones are sufficient.

Comment (dlibenzi) [reply torch_xla/csrc/aten_xla_type.cpp]: Wait ... can't we autogenerate them with `_FN_OUT` in `gen.py`?
I forgot about that 😁

Comment (ailzhang) [reply torch_xla/csrc/aten_xla_type.cpp]: Oh! Nice catch!
